File: server/__synctest.py
#!/usr/bin/env python3
"""
The sync store: what it keeps, what it refuses, and what it hands back.

Run against a temporary database so it never touches a real one.
"""
import base64, json, os, sqlite3, sys, tempfile, threading, urllib.error, urllib.request
from http.server import ThreadingHTTPServer

os.environ["SYNC_USER"] = "cuber"
os.environ["SYNC_PASSWORD"] = "hunter2"
_tmp = tempfile.mkdtemp()
os.environ["SYNC_DB"] = os.path.join(_tmp, "t.db")

sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))
import sync  # noqa: E402

fails = 0
def check(name, cond, extra=""):
    global fails
    if cond:
        print("ok   " + name)
    else:
        fails += 1
        print("FAIL " + name + (("  <" + str(extra) + ">") if extra else ""))

sync.DB = sync.connect()
srv = ThreadingHTTPServer(("127.0.0.1", 0), sync.Handler)
port = srv.server_address[1]
threading.Thread(target=srv.serve_forever, daemon=True).start()
BASE = f"http://127.0.0.1:{port}"
AUTH = "Basic " + base64.b64encode(b"cuber:hunter2").decode()

def call(method, path, body=None, auth=AUTH):
    data = json.dumps(body).encode() if body is not None else None
    req = urllib.request.Request(BASE + path, data=data, method=method)
    if auth:
        req.add_header("Authorization", auth)
    if data:
        req.add_header("Content-Type", "application/json")
    try:
        with urllib.request.urlopen(req) as r:
            return r.status, json.loads(r.read() or b"{}")
    except urllib.error.HTTPError as e:
        return e.code, (json.loads(e.read() or b"{}") if e.headers.get("Content-Type") == "application/json" else {})

rec = lambda tbl, uid, at, data=None, deleted=0: {
    "table": tbl, "uid": uid, "updatedAt": at, "deleted": deleted, "data": data or {"x": uid},
}

# ---- Nothing without the password ----
check("pull without auth is refused", call("GET", "/pull?since=0", auth=None)[0] == 401)
check("push without auth is refused", call("POST", "/push", {"records": []}, auth=None)[0] == 401)
check("a wrong password is refused",
      call("GET", "/pull?since=0", auth="Basic " + base64.b64encode(b"cuber:wrong").decode())[0] == 401)
check("a wrong user is refused",
      call("GET", "/pull?since=0", auth="Basic " + base64.b64encode(b"someone:hunter2").decode())[0] == 401)
check("rubbish in the header is refused", call("GET", "/pull?since=0", auth="Bearer nope")[0] == 401)

# ---- Round trip ----
st, r1 = call("POST", "/push", {"records": [rec("solves", "a", 100), rec("sessions", "s1", 100)]})
check("a push is accepted", st == 200 and r1["applied"] == 2, r1)
st, p = call("GET", "/pull?since=0")
check("everything comes back", st == 200 and len(p["records"]) == 2, p)
check("and the revision with it", p["rev"] == r1["rev"], p)
check("pulling from that revision returns nothing", call("GET", f"/pull?since={p['rev']}")[1]["records"] == [])

# ---- Newer wins; older is ignored ----
call("POST", "/push", {"records": [rec("solves", "a", 200, {"x": "newer"})]})
got = {x["uid"]: x for x in call("GET", "/pull?since=0")[1]["records"]}
check("a newer write replaces an older one", got["a"]["data"] == {"x": "newer"}, got["a"])
st, r = call("POST", "/push", {"records": [rec("solves", "a", 150, {"x": "stale"})]})
check("an older write is ignored", r["applied"] == 0, r)
got = {x["uid"]: x for x in call("GET", "/pull?since=0")[1]["records"]}
check("and does not overwrite", got["a"]["data"] == {"x": "newer"})

# ---- Re-pushing the same thing does not churn the revision ----
before = call("GET", "/pull?since=0")[1]["rev"]
call("POST", "/push", {"records": [rec("solves", "a", 200, {"x": "newer"})]})
after = call("GET", "/pull?since=0")[1]["rev"]
check("re-pushing an unchanged record does not move the revision", before == after, f"{before} -> {after}")

# ---- Deleting is a tombstone, not a hole ----
call("POST", "/push", {"records": [rec("solves", "a", 300, {}, deleted=1)]})
got = {x["uid"]: x for x in call("GET", "/pull?since=0")[1]["records"]}
check("a delete is kept as a tombstone", got["a"]["deleted"] == 1)
check("so the other device is told about it", "a" in got)

# ---- Only the tables that are meant to be there ----
st, r = call("POST", "/push", {"records": [rec("secrets", "x", 400)]})
check("an unknown table is refused", r["applied"] == 0, r)
check("and is not stored", all(x["table"] != "secrets" for x in call("GET", "/pull?since=0")[1]["records"]))
st, r = call("POST", "/push", {"records": [{"table": "solves", "updatedAt": 1}]})
check("a record with no uid is refused", r["applied"] == 0, r)

# ---- One push arrives whole or not at all ----
call("POST", "/push", {"records": [rec("sessions", "s2", 500), rec("solves", "b", 500)]})
revs = {x["uid"]: x["rev"] for x in call("GET", "/pull?since=0")[1]["records"] if x["uid"] in ("s2", "b")}
check("everything in one push shares a revision", len(set(revs.values())) == 1, revs)

# ---- Incremental ----
mark = call("GET", "/pull?since=0")[1]["rev"]
call("POST", "/push", {"records": [rec("reps", "r1", 600)]})
inc = call("GET", f"/pull?since={mark}")[1]
check("only what is new comes back", [x["uid"] for x in inc["records"]] == ["r1"], inc)

# ---- Rubbish in ----
check("a body that is not json is refused", call("POST", "/push", None)[0] in (400, 413))
st, _ = call("POST", "/push", {"records": "not a list"})
check("records must be a list", st == 400)
check("a bad since is refused", call("GET", "/pull?since=abc")[0] == 400)
check("an unknown path is a 404", call("GET", "/nonsense")[0] == 404)

# ---- Health ----
st, h = call("GET", "/health")
check("health reports what is stored", st == 200 and h["ok"] and h["records"] > 0, h)

srv.shutdown()
print("\nALL PASS" if fails == 0 else f"\n{fails} FAILED")
sys.exit(1 if fails else 0)

File: functions/__synctest.mjs
/**
 * The sync store on D1, driven over real HTTP.
 *
 * Once, to give the local D1 its tables:
 *     npm run sync:schema
 * Then in another terminal:
 *     npm run sync:dev
 * and here:
 *     npm run test:sync
 *
 * `wrangler.toml` ships with a placeholder database id, which the local
 * emulator will not accept. Put any UUID there while testing locally, or the
 * real one once `wrangler d1 create` has printed it.
 *
 * Everything here is a promise the client depends on: that a wrong password is
 * refused, that a re-push does not churn the revision, that a newer copy wins
 * and an older one loses, and that deleting keeps a tombstone rather than
 * removing a row.
 */
const BASE = process.env.SYNC_BASE ?? 'http://127.0.0.1:8790/sync';
const auth = (u = 'rogo', p = 'hunter2') => 'Basic ' + Buffer.from(`${u}:${p}`).toString('base64');

let fails = 0;
const check = (name, ok, extra = '') => {
  if (ok) console.log('ok   ' + name);
  else { fails++; console.log('FAIL ' + name + (extra ? `  <${extra}>` : '')); }
};

const call = async (path, { body, user, pass, noAuth } = {}) => {
  const headers = {};
  if (!noAuth) headers.Authorization = auth(user, pass);
  if (body !== undefined) headers['Content-Type'] = 'application/json';
  const res = await fetch(BASE + path, {
    method: body === undefined ? 'GET' : 'POST',
    headers,
    body: body === undefined ? undefined : JSON.stringify(body),
  });
  let json = null;
  try { json = await res.json(); } catch { /* not every refusal has a body */ }
  return { status: res.status, json };
};

const rec = (table, uid, updatedAt, data = {}, deleted = 0) => ({ table, uid, updatedAt, deleted, data });

/* ---- Nothing without the password ---- */
check('no credentials is refused', (await call('/health', { noAuth: true })).status === 401);
check('a wrong password is refused', (await call('/health', { pass: 'wrong' })).status === 401);
check('a wrong user is refused', (await call('/health', { user: 'nope' })).status === 401);
check('the right ones get in', (await call('/health')).status === 200);

/* ---- Pushing, and the revision moving once per push ---- */
const start = (await call('/health')).json.rev;
const one = await call('/push', { body: { records: [
  rec('sessions', 's-1', 1000, { name: 'Main' }),
  rec('solves', 'v-1', 1001, { timeMs: 9870 }),
] } });
check('a push is accepted', one.status === 200 && one.json.applied === 2, JSON.stringify(one.json));
check('and moves the revision exactly once', one.json.rev === start + 1, `${start} -> ${one.json.rev}`);

/* ---- Pulling back what went in ---- */
const pulled = await call(`/pull?since=${start}`);
check('both records come back', pulled.json.records.length === 2, JSON.stringify(pulled.json.records?.length));
check('the session is there with its data',
  pulled.json.records.some((r) => r.uid === 's-1' && r.data.name === 'Main'));
check('the referenced table is written first',
  pulled.json.records[0].table === 'sessions', pulled.json.records[0]?.table);

/* ---- A re-push must not churn ---- */
const again = await call('/push', { body: { records: [rec('sessions', 's-1', 1000, { name: 'Main' })] } });
check('re-pushing the same thing applies nothing', again.json.applied === 0, JSON.stringify(again.json));
check('an older copy loses',
  (await call('/push', { body: { records: [rec('sessions', 's-1', 500, { name: 'Stale' })] } })).json.applied === 0);
const newest = await call('/push', { body: { records: [rec('sessions', 's-1', 2000, { name: 'Renamed' })] } });
check('a newer copy wins', newest.json.applied === 1, JSON.stringify(newest.json));
check('and the new name is what comes back',
  (await call('/pull?since=0')).json.records.find((r) => r.uid === 's-1').data.name === 'Renamed');

/* ---- Only the four tables ---- */
const junk = await call('/push', { body: { records: [rec('secrets', 'x', 9999, { a: 1 })] } });
check('an unknown table is refused', junk.json.applied === 0, JSON.stringify(junk.json));
check('and is not stored',
  (await call('/pull?since=0')).json.records.every((r) => r.table !== 'secrets'));

/* ---- Deleting keeps a tombstone ---- */
await call('/push', { body: { records: [rec('solves', 'v-1', 3000, {}, 1)] } });
const stone = (await call('/pull?since=0')).json.records.find((r) => r.uid === 'v-1');
check('a deleted record stays, marked', stone && stone.deleted === 1, JSON.stringify(stone));

/* ---- Rubbish in ---- */
check('a non-list of records is refused', (await call('/push', { body: { records: 'nope' } })).status === 400);
check('a bad since is refused', (await call('/pull?since=abc')).status === 400);
check('an empty push is harmless', (await call('/push', { body: { records: [] } })).json.applied === 0);

console.log(fails === 0 ? '\nALL PASS' : `\n${fails} FAILED`);
process.exit(fails ? 1 : 0);
